Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -2461,6 +2461,12 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
     if (completedURL.protocolIsJavaScript() && frameElement() && !frameElement()->document().contentSecurityPolicy()->allowJavaScriptURLs(aboutBlankURL().string(), { }, completedURL.string(), frameElement()))
         return;
 
+    RefPtr parent = frame->tree().parent();
+    // If the loader for activeWindow's frame (browsing context) has no outgoing referrer, set its outgoing referrer
+    // to the URL of its parent frame's Document.
+    if (RefPtr activeFrame = activeWindow.frame(); activeFrame && activeFrame->loader().outgoingReferrer().isEmpty() && parent)
+        activeFrame->loader().setOutgoingReferrer(document()->completeURL(dynamicDowncast<LocalFrame>(parent)->document()->url().strippedForUseAsReferrer()));
```

Comment:
Thanks for diligently looking into this.

If the parent is not a LocalFrame, we don't want to crash.  This would happen if the referrer is empty and the iframe is cross-site and site-isolation (currently an experimental feature) is turned on.  Please change to something like this before merging:

RefPtr localParent = dynamicDowncast<LocalFrame>(frame->tree().parent());
if (... && localParent)
    ...completeURL(localParent->document()...)